File: 543-diameter-of-binary-tree/diameter-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
    @maxi = -1
    def depth(node)
        return 0 unless node
        ld = depth(node.left)
        rd = depth(node.right)
        @maxi = [@maxi, ld + rd + 1].max
        return [ld,rd].max + 1
    end
    depth(root)
    return @maxi - 1
end

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxi = 0
        def rec(root):
            nonlocal maxi
            if not root:
                return 0
            left = rec(root.left)
            right = rec(root.right)
            maxi = max(maxi, left + right)
            return max(left, right) + 1
        rec(root)
        return maxi
            
        